src/Middleware/HttpLogging/src/BufferingStream.cs

```diff
@@ -0,0 +1,313 @@
+                {
+                    if (completed)
+                    {
+                        return new string(bufferWriter.WrittenSpan);
+                    }
+                    else
+                    {
+                        EncodingExtensions.Convert(decoder, ReadOnlySequence<byte>.Empty, bufferWriter, flush: true, out charUsed, out completed);
+                    }
+                }
+            }
+            catch (DecoderFallbackException ex)
+            {
+                _logger.DecodeFailure(ex);
+                return "<Decoder failure>";
+            }
+            finally
+            {
+                Reset();
+            }
+        }
+
+        public void Advance(int bytes)
+        {
+            if ((uint)bytes > (uint)_tailMemory.Length)
+            {
+                ThrowArgumentOutOfRangeException(nameof(bytes));
+            }
+
+            _tailBytesBuffered += bytes;
+            _bytesBuffered += bytes;
+            _tailMemory = _tailMemory.Slice(bytes);
+        }
+
+        public Memory<byte> GetMemory(int sizeHint = 0)
+        {
+            AllocateMemory(sizeHint);
+            return _tailMemory;
+        }
+
+        public Span<byte> GetSpan(int sizeHint = 0)
+        {
+            AllocateMemory(sizeHint);
+            return _tailMemory.Span;
+        }
+
+        private void AllocateMemory(int sizeHint)
+        {
+            if (_head == null)
+            {
+                // We need to allocate memory to write since nobody has written before
+                BufferSegment newSegment = AllocateSegment(sizeHint);
+
+                // Set all the pointers
+                _head = _tail = newSegment;
+                _tailBytesBuffered = 0;
+            }
+            else
+            {
+                int bytesLeftInBuffer = _tailMemory.Length;
+
+                if (bytesLeftInBuffer == 0 || bytesLeftInBuffer < sizeHint)
+                {
+                    Debug.Assert(_tail != null);
+
+                    if (_tailBytesBuffered > 0)
+                    {
+                        // Flush buffered data to the segment
+                        _tail.End += _tailBytesBuffered;
+                        _tailBytesBuffered = 0;
+                    }
+
+                    BufferSegment newSegment = AllocateSegment(sizeHint);
+
+                    _tail.SetNext(newSegment);
+                    _tail = newSegment;
+                }
+            }
+        }
+
+        private BufferSegment AllocateSegment(int sizeHint)
+        {
+            BufferSegment newSegment = CreateSegment();
+
+            // We can't use the recommended pool so use the ArrayPool
+            newSegment.SetOwnedMemory(ArrayPool<byte>.Shared.Rent(GetSegmentSize(sizeHint)));
+
+            _tailMemory = newSegment.AvailableMemory;
+
+            return newSegment;
+        }
+
+        private BufferSegment CreateSegment()
+        {
+            return new BufferSegment();
```

Comment:
Why didn't we use a Pipe for this? That way we'd get BufferSegment pooling and would be running on more battle-tested buffering logic.

Reply:
I was using ConcurrentPipeWriter as the base here as it's prior art that does the same thing I want to do: https://github.com/dotnet/aspnetcore/blob/main/src/Servers/Kestrel/Core/src/Internal/Infrastructure/PipeWriterHelpers/ConcurrentPipeWriter.cs. I could have used a Pipe probably, but this is also "battle tested" in a sense as it's used in Kestrel.

Reply:
It as copied from Pipe into Kestrel 😆

Reply:
Yea, now I'm thinking this could just use pipe

Reply:
Then why can't every place we do this pattern be a Pipe as well?

Reply:
ConcurrentPipeWriter's requirements were unique. We needed a PipeWriter that would allow things the DefaultPipeWriter doesn't. Namely appending more data before the previous flush Task completed. This should no longer be necessary after https://github.com/dotnet/aspnetcore/issues/30235 is addressed. If we could have used a default Pipe, we would have.

Reply:
Also, ConcurrentPipeWriter *tries* to be as passthrough as possible to the underlying Pipe. It only starts buffering to avoid overlapping writes on the underlying Pipe and stops buffering ASAP. I would like to remove it after #30235 is addressed, and we could remove it for the HTTP/2 connection.

We might still need it for the response body streams because I expect there's a lot of code out there that works today that doesn't await writes so it'd be a pretty hard break.

Reply:
I'm going to file a follow up issue to consider using a Pipe here instead.